### PR TITLE
CDAP-14905 add a kubernetes twill runner

### PR DESCRIPTION
**Pull request by albertshau**
Add a kubernetes implementation of TwillRunner to support running
programs on Kubernetes instead of YARN.

### CODE REVIEW

**Comment by albertshau** [comment]
https://builds.cask.co/browse/CDAP-DUT6872

File: cdap-kubernetes/src/main/java/co/cask/cdap/k8s/runtime/KubeTwillController.java [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+  }
+
+  @Override
+  public void addLogHandler(LogHandler handler) {
+    // no-op
+  }
+
+  @Override
+  public ServiceDiscovered discoverService(String serviceName) {
+    throw new UnsupportedOperationException("discover service is not supported");
```
Comment (chtyim) [inline cdap-kubernetes/src/main/java/co/cask/cdap/k8s/runtime/KubeTwillController.java]: Guess we don't need this. But we can support it via the KubeDiscoveryService?

File: cdap-kubernetes/src/main/java/co/cask/cdap/k8s/runtime/KubeTwillController.java [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+  }
+
+  @Override
+  public Future<? extends ServiceController> terminate() {
+    if (!completion.isDone()) {
+      // delete the deployment, then delete the config map
+      try {
+        V1DeleteOptions deleteOptions = new V1DeleteOptions();
+        AppsV1Api appsApi = new AppsV1Api(apiClient);
+        appsApi.deleteNamespacedDeployment(name, kubeNamespace, deleteOptions, null, null, null, null);
```
Comment (chtyim) [inline cdap-kubernetes/src/main/java/co/cask/cdap/k8s/runtime/KubeTwillController.java]: Better implement this method using the `async` variant of the delete method and complete the future from the callback (actually the first callback will trigger the async deletion of config map, which triggers the completion of termination future).

File: cdap-kubernetes/src/main/java/co/cask/cdap/k8s/runtime/KubeTwillController.java [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+import java.util.concurrent.Executor;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import javax.annotation.Nullable;
+
+/**
+ * Kubernetes version of a TwillController.
+ */
+public class KubeTwillController implements TwillController {
```
Comment (chtyim) [inline cdap-kubernetes/src/main/java/co/cask/cdap/k8s/runtime/KubeTwillController.java]: extends from `AbstractExecutionServiceController` will make handling of lifecycle events easier. See `NoopTwillController` as an example.

**Review by chtyim — COMMENTED** [review]
Some high level comments

File: cdap-kubernetes/src/main/java/co/cask/cdap/k8s/runtime/KubeTwillController.java (hunk 3 above)
Comment (albertshau) [reply cdap-kubernetes/src/main/java/co/cask/cdap/k8s/runtime/KubeTwillController.java]: That class is in twill-core and uses several guava things. Do you mean copying a guava-less version into master-spi?

Comment (chtyim) [reply cdap-kubernetes/src/main/java/co/cask/cdap/k8s/runtime/KubeTwillController.java]: I see. Nevermind then.

File: cdap-kubernetes/src/main/java/co/cask/cdap/k8s/runtime/KubeTwillController.java [hunk 4]
```diff
@@ -0,0 +1,297 @@
...
+        completion.completeExceptionally(e);
+      }
+    }
+    return completion;
+  }
+
+  @Override
+  public void kill() {
+    try {
+      terminate().get();
```
Comment (chtyim) [inline cdap-kubernetes/src/main/java/co/cask/cdap/k8s/runtime/KubeTwillController.java]: Kill method is not required to block. The contract is only to request for killing it, hence you just need to call terminate() without get.

File: cdap-kubernetes/src/main/java/co/cask/cdap/k8s/runtime/KubeTwillPreparer.java [hunk 5]
```diff
@@ -0,0 +1,476 @@
...
+ * Kubernetes version of a TwillRunner.
+ *
+ * Runs a program in Kubernetes by creating a config-map of the app spec and program options resources that are
+ * expected to be found in the local files of the RuntimeSpecification for the TwillRunnable.
+ * A deployment is created that mounts the created config-map.
+ *
+ * Most of these operations are no-ops as many of these methods and pretty closely coupled to the Hadoop implementation
+ * and have no analogy in Kubernetes.
+ */
+public class KubeTwillPreparer implements TwillPreparer {
```
Comment (chtyim) [inline cdap-kubernetes/src/main/java/co/cask/cdap/k8s/runtime/KubeTwillPreparer.java]: Better be package private class, as it should only be used by the KubeTwillRunner only

File: cdap-kubernetes/src/main/java/co/cask/cdap/k8s/runtime/KubeTwillPreparer.java [hunk 6]
```diff
@@ -0,0 +1,476 @@
...
+    this.runnables = spec.getRunnables().keySet();
+    this.resources = new ArrayList<>();
+    this.discoveryServiceClient = discoveryServiceClient;
+    this.environments = new HashMap<>();
+    for (String runnable : runnables) {
+      environments.put(runnable, new HashMap<>());
+    }
+    this.resourceMeta = resourceMeta;
+    // only expect one runnable for now
+    if (spec.getRunnables().size() != 1) {
```
Comment (chtyim) [inline cdap-kubernetes/src/main/java/co/cask/cdap/k8s/runtime/KubeTwillPreparer.java]: Should be the first thing to check before doing anything.

File: cdap-kubernetes/src/main/java/co/cask/cdap/k8s/runtime/KubeTwillPreparer.java [hunk 7]
```diff
@@ -0,0 +1,474 @@
...
+    labelSelector.matchLabels(resourceMeta.getLabels());
+    spec.setSelector(labelSelector);
+    // TODO: figure out where number of instances is specified in twill spec
+    spec.setReplicas(1);
+
+    /*
+       template:
+         metadata:
+           labels:
+             app: cdap
```
Comment (chtyim) [inline cdap-kubernetes/src/main/java/co/cask/cdap/k8s/runtime/KubeTwillPreparer.java]: For labels, try to follow the operator standard, `cdap.*`. This is also the prefix filtered by the K8s Environment implementation.

File: cdap-kubernetes/src/main/java/co/cask/cdap/k8s/runtime/KubeTwillPreparer.java [hunk 8]
```diff
@@ -0,0 +1,474 @@
...
+         [deployment spec]
+       template:
+         [template spec]
+     */
+    AppsV1Api appsApi = new AppsV1Api(apiClient);
+    V1Deployment deployment = new V1Deployment();
+
+    /*
+       metadata:
+         name: service-runid
```
Comment (chtyim) [inline cdap-kubernetes/src/main/java/co/cask/cdap/k8s/runtime/KubeTwillPreparer.java]: Better prefix the deployment name with `cdap-<instanceName>`, where the `instanceName` is already grabbed by the `KubeMasterEnvironment`.

File: cdap-kubernetes/src/main/java/co/cask/cdap/k8s/runtime/KubeTwillPreparer.java [hunk 9]
```diff
@@ -0,0 +1,474 @@
...
+      .map(entry -> String.format("%s=\"%s\"", entry.getKey(), entry.getValue()))
+      .collect(Collectors.joining("\n"));
+    configMap.putDataItem(podInfo.getLabelsFile(), labelsStr);
+    if (podInfo.getName() != null) {
+      configMap.putDataItem(podInfo.getNameFile(), podInfo.getName());
+    }
+    LOG.trace("Creating config-map {} in Kubernetes", resourceMeta.getName());
+    configMap = api.createNamespacedConfigMap(kubeNamespace, configMap, "true");
+    LOG.trace("Created config-map {} in Kubernetes", resourceMeta.getName());
+    return configMap;
```
Comment (chtyim) [inline cdap-kubernetes/src/main/java/co/cask/cdap/k8s/runtime/KubeTwillPreparer.java:520]: Set owner references to the config map as well.

File: cdap-kubernetes/src/main/java/co/cask/cdap/master/environment/k8s/PodInfo.java [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+    this.dir = dir;
+    this.labelsFile = labelsFile;
+    this.nameFile = nameFile;
+    this.namespace = namespace;
+    this.name = name;
+    this.labels = labels;
+    this.ownerReferences = ownerReferences;
+  }
+
+  public String getDir() {
```
Comment (chtyim) [inline cdap-kubernetes/src/main/java/co/cask/cdap/master/environment/k8s/PodInfo.java]: What dir is it? What is it for?

File: cdap-kubernetes/src/main/java/co/cask/cdap/master/environment/k8s/PodInfo.java [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+  public String getNameFile() {
+    return nameFile;
+  }
+
+  public String getNamespace() {
+    return namespace;
+  }
+
+  @Nullable
+  public String getName() {
```
Comment (chtyim) [inline cdap-kubernetes/src/main/java/co/cask/cdap/master/environment/k8s/PodInfo.java]: What is this name? The CDAP name in the K8s deployment?

File: cdap-kubernetes/src/main/java/co/cask/cdap/k8s/runtime/KubeTwillPreparer.java (hunk 7 above)
Comment (albertshau) [reply cdap-kubernetes/src/main/java/co/cask/cdap/k8s/runtime/KubeTwillPreparer.java]: these are old comments that don't reflect reality, will update.

File: cdap-kubernetes/src/main/java/co/cask/cdap/master/environment/k8s/PodInfo.java (hunk 10 above)
Comment (albertshau) [reply cdap-kubernetes/src/main/java/co/cask/cdap/master/environment/k8s/PodInfo.java]: will rename to podInfoDir. This is used as the mount directory for the config-map, which contains the pod labels, etc.

File: cdap-kubernetes/src/main/java/co/cask/cdap/master/environment/k8s/PodInfo.java (hunk 11 above)
Comment (albertshau) [reply cdap-kubernetes/src/main/java/co/cask/cdap/master/environment/k8s/PodInfo.java]: yeah, will rename it to cdapInstanceName

File: cdap-kubernetes/src/main/java/co/cask/cdap/k8s/runtime/KubeTwillPreparer.java (hunk 9 above)
Comment (albertshau) [reply cdap-kubernetes/src/main/java/co/cask/cdap/k8s/runtime/KubeTwillPreparer.java:520]: they are set. Every resource here will share the same meta object, just will have a different kind.

File: cdap-kubernetes/src/main/java/co/cask/cdap/master/environment/k8s/PodInfo.java (hunk 11 above)
Comment (albertshau) [reply cdap-kubernetes/src/main/java/co/cask/cdap/master/environment/k8s/PodInfo.java]: actually, no this is the pod name that is used to get the owner references

File: cdap-kubernetes/src/main/java/co/cask/cdap/k8s/runtime/KubeTwillPreparer.java [hunk 12]
```diff
@@ -0,0 +1,522 @@
...
+    this.resources = new ArrayList<>();
+    this.discoveryServiceClient = discoveryServiceClient;
+    this.environments = new HashMap<>();
+    for (String runnable : runnables) {
+      environments.put(runnable, new HashMap<>());
+    }
+    this.resourceMeta = resourceMeta;
+    RuntimeSpecification runtimeSpecification = spec.getRunnables().values().iterator().next();
+    ResourceSpecification resourceSpecification = runtimeSpecification.getResourceSpecification();
+    appSpec = runtimeSpecification.getLocalFiles().stream()
```
Comment (chtyim) [inline cdap-kubernetes/src/main/java/co/cask/cdap/k8s/runtime/KubeTwillPreparer.java]: To be consistent with other fields in the constructor, should be `this.appSpec`.

File: cdap-kubernetes/src/main/java/co/cask/cdap/k8s/runtime/KubeTwillPreparer.java [hunk 13]
```diff
@@ -0,0 +1,522 @@
...
+    deployment = appsApi.createNamespacedDeployment(kubeNamespace, deployment, "true");
+    LOG.trace("Creating deployment {} in Kubernetes", resourceMeta.getName());
+    return deployment;
+  }
+
+  /*
+           volumes:
+           - name: config-volume
+             configMap:
+               name: service-runid
```
Comment (chtyim) [inline cdap-kubernetes/src/main/java/co/cask/cdap/k8s/runtime/KubeTwillPreparer.java]: The name is not service-runid, right? and it actually contains appSpec and programOpts inside, right?

File: cdap-kubernetes/src/main/java/co/cask/cdap/master/environment/k8s/KubeMasterEnvironment.java [hunk 14]
```diff
@@ -157,25 +157,53 @@ public String getName() {
...
-   */
-  private List<V1OwnerReference> getOwnerReference(String podNamePath, String namespace) {
+  private PodInfo getPodInfo(Map<String, String> conf) throws IOException {
+    String namespace = conf.getOrDefault(NAMESPACE_KEY, DEFAULT_NAMESPACE);
+
+    File podInfoDir = new File(conf.getOrDefault(POD_INFO_DIR, DEFAULT_POD_INFO_DIR));
+    if (!podInfoDir.isDirectory()) {
+      throw new IllegalArgumentException(String.format("%s is not a directory.", podInfoDir.getAbsolutePath()));
+    }
+
```
Comment (chtyim) [inline cdap-kubernetes/src/main/java/co/cask/cdap/master/environment/k8s/KubeMasterEnvironment.java:177]: Remove extra new line

**Review by chtyim — APPROVED** [review]
Just some minor formatting comments.

File: cdap-kubernetes/src/main/java/co/cask/cdap/k8s/runtime/KubeTwillPreparer.java (hunk 12 above)
Comment (albertshau) [reply cdap-kubernetes/src/main/java/co/cask/cdap/k8s/runtime/KubeTwillPreparer.java]: done

File: cdap-kubernetes/src/main/java/co/cask/cdap/k8s/runtime/KubeTwillPreparer.java (hunk 13 above)
Comment (albertshau) [reply cdap-kubernetes/src/main/java/co/cask/cdap/k8s/runtime/KubeTwillPreparer.java]: right, fixed

**Pull request merged**